src/lib/desmo-module.ts

```diff
@@ -179,33 +179,42 @@ export class Desmo {
     taskId: string;
     result: string;
   }> {
-    const myIexec = this.iexec as IExec;
+    if (this.iexec === undefined) {
+      throw new Error('A connection to iExec is required!');
+    }
     if (!this.isConnected) {
       throw new Error(
         'This method requires the wallet signer to be already signed-in!',
       );
     }
-    
-    // eslint-disable-next-line @typescript-eslint/no-this-alias
-    const self = this;
-    return new Promise( (resolve, reject) => {
-      (function loop(): void {
-        setTimeout(async () => {
-          const taskId = await self.retrieveTaskID();
-          const taskDetail = await myIexec.task.show(taskId);
-          console.log(`The status of the task ${taskId} is ${taskDetail.statusName}`);
-          if (taskDetail.statusName === TaskStatus.COMPLETED) {
-            const tx = await self.contract.receiveResult(taskId, '0x00');
-            await tx.wait();
-            const result = await self.contract.getQueryResult(taskId);
-            // TODO: Decode the result
-            resolve({ requestId: '', taskId, result });
-          } else {
-            loop();
-          }
-        }, 1000)
-      })();
+
+    const taskId = await this.retrieveTaskID();
+
+    const taskObservable = await this.iexec.task.obsTask(taskId, {
+      dealid: this.dealId,
     });
+
+    const taskCompletion: Promise<ethers.Bytes> = new Promise(
+      (resolve, reject) => {
+        taskObservable.subscribe({
+          next: ({ message }) =>
+            console.log(`The status of the task ${taskId} is ${message}.`),
+          error: (e) => reject(e),
+          complete: async () => {
+            const tx = await this.contract.receiveResult(taskId, '0x00');
+            await tx.wait();
+            const result: ethers.Bytes = await this.contract.getQueryResult(
+              taskId,
+            );
+            resolve(result);
```

Comment:
Are you sure that in the `complete` callback, the status of the task is always `TaskStatus.COMPLETED?` What happens if the Task fails? In some libraries, the error callback is only called when some network error happens. What does the iExec documentation say about this?